### PR TITLE
Change debounce export type

### PR DESCRIPTION
**Pull request by faergeek**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

TL;DR: This is similar to #21062

Right now I use `awesome-type-script-loader` for webpack and it complains like this:
```
TS2497: Module '"<...>/node_modules/@types/debounce/index"' resolves to a non-module entity and cannot be imported using this construct.
```

I import it like this:
```
import * as debounce from 'debounce'
```

### CODE REVIEW

**Comment by faergeek** [comment]
tslint complains exactly about this change. I'm not sure why that rule is good or not, but with function declaration `awesome-typescript-loader` complains as I wrote above.

**Comment by denis-sokolov** [comment]
I don’t understand why the change is functional at all. Seems to me like it should be identical whether pre-PR or post-PR.

**Comment by faergeek** [comment]
It seems that it doesn't allow `export = ...` to contain function declaration. Actually in real code you cannot assign function declaration to `module.exports`, for example, it will be treated as function expression. Just thoughts, not sure 100%.

**Comment by denis-sokolov** [comment]
It is my understanding that function declarations and function expressions are types of syntax, but these are not different types semantically and they are not different at runtime. The type of the exported object is a function with the described parameters and return values, regardless of how it was created in the original source code.

I trust you have tested it well that the new syntax works whereas the previous doesn’t. But I would fix it systematically, rather than manually adjust all typings. Perhaps it is possible to fix it in the TypeScript compiler. If it is a meaningful distinction in the TypeScript compiler, and we do, indeed, need to change all the typings, then perhaps `prefer-declare-function` should be disabled globally, and not on a module-per-module basis, creating an impression of a hack.

Please don’t misunderstand. My intention is not to criticize the improvement of the typings, rather, that I am uncomfortable with individual changes for individual typings. To top it off, I am not a maintainer, so my opinion is of limited value. :-)

**Comment by Bnaya** [comment]
FYI
My fix for that issue is:
```ts
declare module "debounce" {
  function debounce<F extends Function>(func: F, wait?: number, immediate?: boolean): F;
  namespace debounce {

  }
  export = debounce;
}

```

**Comment by RyanCavanaugh** [comment]
This PR has been open and unchanged 5 days without signoff or complaint. This will be merged by a maintainer soon if there are no objections.

**Comment by DanielRosenwasser** [comment]
Sorry, I don't understand why this change is being made, even after the above conversation. It sounds like it's not stylistic though. Could you explain?

**Comment by faergeek** [comment]
Sorry, I can't explain because I'm newbie in TypeScript. It just works for me only when written like that.

**Comment by DanielRosenwasser** [comment]
Sure, let's work through this together. How does it work for you any differently than currently? How are you trying to use it right now?

**Comment by faergeek** [comment]
Here's a minimal case https://github.com/faergeek/debounce-ts-error

```sh
yarn install
yarn start
```

**Comment by DanielRosenwasser** [comment]
I believe that's because you're trying to use an `import *` style import, which shouldn't allow you to call debounce (it's a typescript bug which is too prevalent to fix now). The correct way to import is with

```ts
import debounce = require("debounce");
```

We'll have a solution for users in the TypeScript 2.7 timeframe.

**Comment by faergeek** [comment]
When imported like you suggested it complains:

```
Import assignment cannot be used when targeting ECMAScript 2015 modules. Consider using 'import * as ns from "mod"', 'import {a} from "mod"', 'import d from "mod"', or another module format instead
```

I understand that I can change module target in config, etc., but actually I need to target `es2015` modules for various babel and webpack features.

**Comment by Bnaya** [comment]
Hey @faergeek,
I tried your code, and it dose not work for me
I'm still getting the same error.

```ts
declare module "debounce" {
  function debounce<F extends Function>(func: F, wait?: number, immediate?: boolean): F;
  namespace debounce {

  }
  export = debounce;
}

```
dose work for me

**Comment by DanielRosenwasser** [comment]
If you're using Babel/Webpack to transform your output, you *should be able to just run with `allowSyntheticDefaultImports` and using a default-import instead.

**Comment by faergeek** [comment]
Shouldn't types work with any TypeScript config?

**Comment by DanielRosenwasser** [comment]
I would like to say yes, but the world of ES/CommonJS interop is problematic.

**Comment by faergeek** [comment]
@DanielRosenwasser could you then please fork this https://github.com/faergeek/debounce-ts-error and make it work?

**Comment by sandersn** [comment]
For housekeeping purposes I am closing this PR since it is not correct. Feel free to continue the conversation here so you can figure out how to make projects that use debounce work.
